lib/active_model/serializer/adapter/json_sideload.rb

```diff
@@ -0,0 +1,185 @@
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonSideload < Json
+
+        # the list of what has already been serialized will be kept here to
+        # help avoid infinite recursion
+        #
+        # this should be a list of association_name to a list of objects
+        attr_accessor :associations
+
+        # When we are sideloading associations, we can more easily track what
+        # has been serialized, so that we avoid infinite
+        # recursion / serialization.
+        def initialize(*args)
+          @associations = {}
+
+          super(*args)
+        end
+
+        def serializable_hash(options = nil)
+          options ||= {}
+
+          serialize_hash(options)
+
+          singularize_lone_associations
+
+          @associations
+        end
+
+
+        def serialize_hash(options)
+          result = nil
+
+          if serializer.respond_to?(:each)
+            #TODO: Is this ever hit?
+            # binding.pry
+            ap "wwwwwwwwwwwwwwwwwwwwww"
+            result = serialize_array(serializer, options)
```

Comment:
when the resource is a collection it uses the array serializer with has `each`, e.g. `render json: User.all` or `render json: [user, post]` would both go in the `each` block.  I'm thinking though, in https://github.com/rails-api/active_model_serializers/pull/1004#discussion_r38288295 it might make sense to have a 'collection' adapter and 'resource' adapter to clean this up and separate concerns.. though it introduces a bit more complexity, so we'd need to add pieces of it whenever it's more useful than not :)